ianalyzer_readers/readers/core.py

```diff
@@ -142,20 +174,172 @@ def source2dicts(self, source: Source) -> Iterable[Document]:
         Given a source file, returns an iterable of extracted documents.
 
         Parameters:
-            source: the source file to extract. This can be a string with the path to
-                the file, or a tuple with a path and a dictionary containing metadata.
-                Some reader subclasses may also support bytes as input.
+            source: Source to extract.
         
         Returns:
             an iterable of document dictionaries. Each of these is a dictionary,
                 where the keys are names of this Reader's `fields`, and the values
                 are based on the extractor of each field.
+        '''
+
+        self.validate()
+
+        data, metadata = self.data_and_metadata_from_source(source)
+
+        if isinstance(data, AbstractContextManager):
+            context_manager = data
+        else:
+            context_manager = nullcontext(data)
```

Comment:
Why use the context manager around `data`? Is there any state that may be lost without it?

Reply:
`data` may be implemented as a context manager so the child class can have some kind of teardown (e.g. closing a file). In these lines, the `nullcontext` manager is added if `data` is not already a context manager; that's to avoid repetition in the steps below.